channeldb/payment_control.go

```diff
@@ -615,19 +619,36 @@ func fetchPaymentBucketUpdate(tx kvdb.RwTx, paymentHash lntypes.Hash) (
 
 // nextPaymentSequence returns the next sequence number to store for a new
 // payment.
-func nextPaymentSequence(tx kvdb.RwTx) ([]byte, error) {
-	payments, err := tx.CreateTopLevelBucket(paymentsRootBucket)
-	if err != nil {
-		return nil, err
-	}
+func (p *PaymentControl) nextPaymentSequence() ([]byte, error) {
+	p.paymentSeqMx.Lock()
+	defer p.paymentSeqMx.Unlock()
+
+	// Set a new upper bound in the DB every 1000 payments to avoid
+	// conflicts on the sequence when using etcd.
+	if p.currPaymentSeq == p.storedPaymentSeq {
+		var newUpperBound uint64
+
+		if err := kvdb.Update(p.db.Backend, func(tx kvdb.RwTx) error {
+			paymentsBucket, err := tx.CreateTopLevelBucket(
+				paymentsRootBucket,
+			)
+			if err != nil {
+				return err
+			}
 
-	seq, err := payments.NextSequence()
-	if err != nil {
-		return nil, err
+			newUpperBound = paymentsBucket.Sequence() + 1000
+			return paymentsBucket.SetSequence(newUpperBound)
+		}, func() {}); err != nil {
+			return nil, err
+		}
+
+		p.storedPaymentSeq = newUpperBound
 	}
 
+	p.currPaymentSeq++
```

Comment:
I think we need to set this value on startup. Otherwise we'll start at 0 when using this code. That'll also make the `if p.currPaymentSeq == p.storedPaymentSeq` not be true on first run. Which probably means we need to bump the upper bound on startup as well.

Reply:
Thanks, that's a good catch I missed. Should be fixed now.